### PR TITLE
uninstall script

### PR DESCRIPTION
**Pull request by xu-cheng**
cc @Homebrew/owners @DomT4 @chdiza

### CODE REVIEW

**Comment by xu-cheng** [comment]
Tweaked a bit. It works without any brew CLI or internal API.

Ask for comments.

File: uninstall [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+  ].map { |p| Pathname.new(p) }
+end
+
+if /darwin/i === RUBY_PLATFORM
+  $HOMEBREW_FILES += %W[
+    /Applications
+    #{ENV["HOME"]}/Applications
+  ].map do |p|
+    Pathname.new(p).subdirs.select do |app|
+      app.resolved_path.to_s.start_with? $HOMEBREW_PREFIX.to_s
```
Comment (tdsmith) [inline uninstall]: It's possible for HOMEBREW_CELLAR to live outside HOMEBREW_PREFIX; I think this should be HOMEBREW_CELLAR

**Comment by tdsmith** [comment]
Elaborating: it's a weird configuration but Homebrew's stuff doesn't always live inside `HOMEBREW_PREFIX`: https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Installation.md#untar-anywhere-and-then-symlink-the-brew-command-elsewhere -- `global.rb` computes the possibilities.

File: uninstall [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+end
+
+####################################################################### script
+
+$HOMEBREW_PREFIX_CANDIDATES=[]
+
+OptionParser.new do |opts|
+  opts.banner = "Homebrew Uninstaller\nUsage: ./uninstall [options]"
+  opts.summary_width = 16
+  opts.on("-pPATH", "--path=PATH", "Manually set path to Homebrew.") { |p| $HOMEBREW_PREFIX_CANDIDATES << Pathname.new(p) }
```
Comment (tdsmith) [inline uninstall]: "Path to Homebrew" is ambiguous; "Homebrew prefix, like /usr/local" might be better

File: uninstall [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+  opts.on("--skip-cache-log", "Skipes removal of Caches and Logs files.") { |p| options[:skip_cache_log] = true }
```
Comment (tdsmith) [inline uninstall]: s/Skipes/Skips/

File: uninstall [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    next
+  end
+
+  if file.symlink?
+    system "unlink", file.to_s
+  elsif file.directory?
+    system "rm", "-rf", file.to_s
+  else
+    system "rm", file.to_s
+  end
```
Comment (jacknagel) [inline uninstall]: `system "unlink"` is the same as `system "rm"` (except unlink doesn't accept flags and only accepts one file argument).

But I think all three branches can be removed in favor of just calling `FileUtils.rm_rf(file)` or equivalent.

**Comment by xu-cheng** [comment]
Sorry about the delay. I have updated to reflect the comments.

**Comment by xu-cheng** [comment]
Any other comment?

**Comment by xu-cheng** [comment]
Ping @mikemcquaid for review.

File: uninstall [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+end
+
+####################################################################### script
+
+$HOMEBREW_PREFIX_CANDIDATES=[]
+
+OptionParser.new do |opts|
+  opts.banner = "Homebrew Uninstaller\nUsage: ./uninstall [options]"
+  opts.summary_width = 16
+  opts.on("-pPATH", "--path=PATH", "Set Homebrew prefix, like /usr/local.") { |p| $HOMEBREW_PREFIX_CANDIDATES << Pathname.new(p) }
```
Comment (mikemcquaid) [inline uninstall]: `Sets HOMEBREW_PREFIX. Defaults to /usr/local.`

File: uninstall [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+  opts.on("--skip-cache-log", "Skips removal of Caches and Logs files.") { |p| options[:skip_cache_log] = true }
```
Comment (mikemcquaid) [inline uninstall]: `Skips removal of HOMEBREW_CACHE` and make it `--skip-cache`.

File: uninstall [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+  opts.on("-f", "--force", "Forces removal of files, bypassing prompt. USE WITH CAUTION!") { options[:force] = true }
```
Comment (mikemcquaid) [inline uninstall]: `Uninstall without prompting.` Think we can kill the `USE WITH CAUTION!`.

File: uninstall [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+  opts.on("-q", "--quiet", "Suppress all output.") { options[:quiet] = true }
+  opts.on("-d", "--dry-run", "Dry run, print results, but perform no actual operations.") { options[:dry_run] = true }
```
Comment (mikemcquaid) [inline uninstall]: `Simulate uninstall but don't remove anything.`

File: uninstall [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+  HOMEBREW_REPOSITORY/"Cellar"
+end
+
+$HOMEBREW_FILES = %W[
+  .git .gitignore .yardopts CODEOFCONDUCT.md CONTRIBUTING.md
+  LICENSE.txt README.md SUPPORTERS.md Library/.rubocop.yml
+  Library/Aliases Library/Contributions Library/ENV Library/Formula
+  Library/Homebrew Library/LinkedKegs Library/Taps Library/brew.rb
+  bin/brew share/doc/homebrew share/man/man1/brew.1
+].map { |p| HOMEBREW_REPOSITORY/p }
```
Comment (mikemcquaid) [inline uninstall]: Might be an idea to download read Homebrew's `gitignore` so this doesn't need to be kept up-to-date?

File: uninstall (hunk 6 above)
Comment (mikemcquaid) [inline uninstall]: Ah, actually, `Skips removal of HOMEBREW_CACHE and HOMEBREW_LOGS` and make it `--skip-cache-and-logs`.

File: uninstall [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+  p = `command -v brew`
+  $HOMEBREW_PREFIX_CANDIDATES << Pathname.new(p.strip).dirname.parent unless p.empty?
+  $HOMEBREW_PREFIX_CANDIDATES << Pathname.new("/usr/local") # Homebrew default path
+  $HOMEBREW_PREFIX_CANDIDATES << Pathname.new("#{ENV["HOME"]}/.linuxbrew") # Linuxbrew default path
+end
+
+HOMEBREW_PREFIX = $HOMEBREW_PREFIX_CANDIDATES.detect do |p|
+  p.directory? && (p/"bin/brew").executable?
+end
+abort "Failed to locate Homebrew" if HOMEBREW_PREFIX.nil?
```
Comment (mikemcquaid) [inline uninstall]: `Failed to locate Homebrew!`

File: uninstall [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+$HOMEBREW_FILES.select!(&:exist?)
+
+unless options[:quiet]
+  warn "This script #{options[:dry_run] ? "would" : "will"} remove:"
+  $HOMEBREW_FILES.each(&:pretty_print)
+end
+
+unless options[:force] || options[:dry_run]
+  STDERR.print "Confirm to uninstall Homebrew? [y/N] "
```
Comment (mikemcquaid) [inline uninstall]: `Are you sure you want to uninstall Homebrew? [y/N]`

File: uninstall [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+  map { |p| HOMEBREW_PREFIX/p }.select(&:exist?).map(&:to_s)
+args = %W[-E] + paths + %W[-regex .*/info/([^.][^/]*\.info|dir)]
+if options[:dry_run]
+  args << "-print"
+else
+  args += %W[-exec /bin/bash -c]
+  args << "/usr/bin/install-info --delete --quiet {} \"$(dirname {})/dir\""
+  args << ";"
+end
+puts "Would uninfo:" if options[:dry_run]
```
Comment (mikemcquaid) [inline uninstall]: uninfo?

File: uninstall [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+  warn "This script #{options[:dry_run] ? "would" : "will"} remove:"
+  $HOMEBREW_FILES.each(&:pretty_print)
+end
+
+unless options[:force] || options[:dry_run]
+  STDERR.print "Confirm to uninstall Homebrew? [y/N] "
+  abort unless gets.rstrip =~ /y|yes/i
+end
+
+ohai "Unlink Homebrew Installation..." unless options[:quiet]
```
Comment (mikemcquaid) [inline uninstall]: `Removing Homebrew installation...`

File: uninstall (hunk 12 above)
Comment (mikemcquaid) [inline uninstall]: Probably just `Would delete:`

File: uninstall [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+end
+puts "Would uninfo:" if options[:dry_run]
+system "/usr/bin/find", *args
+args = paths + %W[-type l -lname */Cellar/*]
+if options[:dry_run]
+  args << "-print"
+else
+  args += %W[-exec unlink {} ;]
+end
+puts "Would unlink:" if options[:dry_run]
```
Comment (mikemcquaid) [inline uninstall]: `Would delete:`

File: uninstall [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+system "/usr/bin/find", *args
+
+ohai "Remove Homebrew Installation..." unless options[:quiet]
+$HOMEBREW_FILES.each do |file|
+  if options[:dry_run]
+    puts "Would remove #{file}"
```
Comment (mikemcquaid) [inline uninstall]: `Would delete:`

File: uninstall [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+  else
+    FileUtils.rm_rf(file)
+  end
+end
+
+ohai "Cleanup..." unless options[:quiet]
```
Comment (mikemcquaid) [inline uninstall]: Not sure what this bit is for? Would be good to elaborate

File: uninstall [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+paths << HOMEBREW_REPOSITORY/"Library"
+paths.select!(&:exist?)
+paths.map!(&:to_s)
+args = paths + %W[-name .DS_Store]
+if options[:dry_run]
+  args << "-print"
+else
+  args << "-delete"
+end
+puts "Would remove:" if options[:dry_run]
```
Comment (mikemcquaid) [inline uninstall]: `Would delete:`

File: uninstall [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+end
+puts "Would remove directories:" if options[:dry_run]
+system "/usr/bin/find", *args
+# remove HOMEBREW_REPOSITORY HOMEBREW_PREFIX if they're empty.
+unless options[:dry_run]
+  Kernel.system "rmdir #{HOMEBREW_REPOSITORY} &>/dev/null"
+  Kernel.system "rmdir #{HOMEBREW_PREFIX} &>/dev/null"
+end
+
+ohai "Uninstallation done!" unless options[:quiet]
```
Comment (mikemcquaid) [inline uninstall]: `Homebrew uninstalled!`

File: uninstall [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+
+residual_files = []
+residual_files.concat(HOMEBREW_REPOSITORY.subdirs) if HOMEBREW_REPOSITORY.exist?
+residual_files.concat(HOMEBREW_PREFIX.subdirs) if HOMEBREW_PREFIX.exist?
+residual_files.uniq!
+
+unless residual_files.empty? || options[:quiet]
+  puts "Following residual files found:"
```
Comment (mikemcquaid) [inline uninstall]: `The following files could not be deleted:`

File: uninstall [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+  residual_files.each(&:pretty_print)
+  puts "You may consider to remove them by yourself.\n"
```
Comment (mikemcquaid) [inline uninstall]: `You may need to use sudo if you wish to delete them yourself`.

File: uninstall [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+end
+
+if /darwin/i === RUBY_PLATFORM && HOMEBREW_PREFIX.exist? &&
+  HOMEBREW_PREFIX.to_s == "/usr/local" && !options[:quiet]
+  puts <<-EOS
+You may consider to restore /usr/local to its original ownership.
```
Comment (mikemcquaid) [inline uninstall]: `You may want to restore /usr/local's original permissions`

File: uninstall (hunk 16 above)
Comment (xu-cheng) [inline uninstall]: This is used to remove all the empty directories.

File: uninstall (hunk 19 above)
Comment (xu-cheng) [inline uninstall]: Technically, most residual files are the files we aren't sure about their original. Such as `HOMEBREW_PREFIX/lib/python2.7` or `HOMEBREW_PREFIX/var/log`.

File: uninstall (hunk 9 above)
Comment (xu-cheng) [inline uninstall]: Good idea. :+1:

File: uninstall (hunk 16 above)
Comment (mikemcquaid) [inline uninstall]: `Removing empty directories...`

File: uninstall (hunk 19 above)
Comment (mikemcquaid) [inline uninstall]: `The following possible Homebrew files were not deleted:`

File: uninstall (hunk 9 above)
Comment (xu-cheng) [inline uninstall]: A little question. Which `gitignore` file should I read? The local one or online one. The local one may be missing, while the online one may not be consistence with local installation.

Comment (mikemcquaid) [inline uninstall]: Local and if it's missing online (and if download fails, error out?)

**Comment by xu-cheng** [comment]
Update.

**Comment by xu-cheng** [comment]
Any more suggestions? OR merge it?

**Comment by mikemcquaid** [comment]
Merge away :+1: (and link to it from the docs)

**Pull request closed without merging**